### PR TITLE
issue #23 - add ability to update some of the block settings for existing blocks

### PR DESCRIPTION
**Pull request by dmitriim**
**This patch:** 
- adds a new feature to update config, visibility, subcontext and page pattern for already existing blocks to fix issue #23
- adds tests to support a new feature
- modifies blocks set up form to make it a bit more clear from UI what fields could be applied for different actions (add, reposition, update). So not relevant fields are becoming disabled when the action that can't use those fields is selected. 
- adds extra tests to cover reposition functionality
- fixes hard-coded "Visible?" lang string 

**Testing.**
This patch adds ability to modify blocks once they are already added. For example a case from https://github.com/catalyst/moodle_tool_blocksmanager/issues/23

0. Have a category with 3 courses. Each course should have at least one assignment activity. 
1. Add Online users block to that category using the block manager plugin using course-view-* pattern and Show in subcontexts set to No. Add new rule, Apply set of blocks and process adhoc tasks. 
2. Confirm that you can see the block on the main course page for each of the courses. 
3. Now you want move that block to assignment pages only.
4. You need to create a new rule set as following

- Block - Online users
- Update instance (if exists)? - Yes
- Show in subcontexts - Yes
- Page type pattern - mod-assign-view

5. Add new rule, Apply set of blocks and process adhoc tasks. 
6. Confirm  that you can't see the block anymore on any of the pages in the courses except assignment pages.

### CODE REVIEW

**Comment by dmitriim** [comment]
Hey! @danmarsden @sarahjcotton If you guys have some time for reviewing this change that would be awesome.

File: classes/form/setup_form.php [hunk 1]
```diff
@@ -61,13 +61,29 @@ protected function definition() {
...
+        $this->_form->addElement('select', 'reposition',  get_string('field_reposition', 'tool_blocksmanager'),
+            [0 => get_string('no'), 1 => get_string('yes')]);
+
+        $this->_form->disabledIf('add', 'reposition', 'eq', 1);
+        $this->_form->disabledIf('add', 'update', 'eq', 1);
+
         $this->_form->disabledIf('reposition', 'add', 'eq', 1);
+        $this->_form->disabledIf('reposition', 'update', 'eq', 1);
+
+        $this->_form->disabledIf('update', 'add', 'eq', 1);
```
Comment (TomoTsuyuki) [inline classes/form/setup_form.php:84]: `disabledIf` functions are around anywhere and hard to understand.
I think it's better to group in one place after same element, or same dependency?
e.g.
`$this->_form->disabledIf('update', 'add', 'eq', 1);`  should be immediate after `update` element, or `add` element.

File: classes/setup_item.php [hunk 2]
```diff
@@ -121,18 +127,19 @@ class setup_item {
...
      *
      *  If reposition = 1
      *
-     *  8 - second region
-     *  9 - second weight
-     *  10 - show in subcontexts
-     *  11 - page type pattern
+     *  9 - second region
+     *  10 - second weight
+     *  12 - show in subcontexts
```
Comment (TomoTsuyuki) [inline classes/setup_item.php:141]: This should be `11` :)

**Review by TomoTsuyuki — COMMENTED** [review]
I tested as the instruction, and it's completed successfully.
Just left few comments in the codes.

File: classes/form/setup_form.php (hunk 1 above)
Comment (dmitriim) [reply classes/form/setup_form.php]: fixed

File: classes/setup_item.php (hunk 2 above)
Comment (dmitriim) [reply classes/setup_item.php]: ah. good catch. Fixed that

**Pull request merged**